brainframe_qt/ui/dialogs/task_configuration/core/zone.py

```diff
@@ -0,0 +1,139 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from typing import List, Optional, Tuple, TypeVar
+
+from brainframe.api import bf_codecs
+from shapely import geometry
+
+T = TypeVar("T")
+
+
+@dataclass
+class Zone(ABC):
+    coords: Optional[List[Tuple[int, int]]]
+    """Coordinates of the endpoints/vertices of the Zone. Top left of frame is (0, 0)
+    
+    The full-frame region has no coords
+    """
+
+    name: Optional[str] = None
+    """The name of the Zone, if it has one"""
+
+    id: Optional[int] = None
+    """The id of the Zone known by the BrainFrame backend"""
+
+    alarms: List[bf_codecs.ZoneAlarm] = field(default_factory=list)
+    """A list of Alarms that the Zone holds"""
+
+    @abstractmethod
+    def copy(self: T) -> T:
+        """Deep copy"""
+
+    @abstractmethod
+    def is_shape_ready(self) -> bool:
+        """Whether this type of Zone has enough points to be considered complete"""
+
+    @abstractmethod
+    def takes_additional_points(self) -> bool:
+        """Whether more points can be added to this type of Zone"""
+
+    @abstractmethod
+    def would_be_valid_zone(self, new_vertex: Tuple[int, int]) -> bool:
+        """Whether the addition of new_vertex to the zone would result in a valid
+        zone"""
+
+    def add_vertex(self, vertex: Tuple[int, int]) -> None:
+        assert self.coords is not None
+        self.coords.append(vertex)
+
+    def adjust_final_vertex(self, vertex: Tuple[int, int]) -> None:
+        assert self.coords is not None
```

Comment:
Should these asserts be here still?

Reply:
These asserts are for type hinting

Reply:
Okay, in that case, why not do:
```
coords: Tuple[int, int]
```

But if you want to raise an error, then do a TypeError instead of an assertion

Reply:
Mypy doesn't allow you to reassign type information like that. While we're not using Mypy _yet_, I'd like to in the future. There are two ways that you can do this in Mypy

```python
assert self.coords is not None
```
```python
self.coords = typing.cast(Tuple[int, int], self.coords)
```

I think the first is cleaner, but I'm open to discussion.

Reply:
Raising a `TypeError` is an interesting idea, but my impression was that `TypeError` is meant to be raised when a function is passed arguments of the wrong type.

> Raised when an operation or function is applied to an object of inappropriate type
> This exception may be raised by user code to indicate that an attempted operation on an object is not supported, and is not meant to be.

(https://docs.python.org/3/library/exceptions.html#TypeError)

I guess that could work here?

Reply:
The issue with the `TypeError` in _this_ situation is that I'm trying to run functions that shouldn't be run on an object with specific _state_ (i.e. the full-frame zone), not specific _type_

Reply:
I went ahead and just used `RuntimeError`s. Let me know what you think

Reply:
Works with me. I just try to keep AssertionError's out of the codebase after I'm done developing a feature.